docs/concepts/validators.md

```diff
@@ -0,0 +1,129 @@
+
+As an example, the `ExtractedSummarySentencesMatch` validator accepts a `filepaths` property in the metadata dictionary to specify what source files to compare the summary against to ensure similarity.  Unlike arguments which are specified at validator initialization, metadata is specified when calling `guard.validate` or `guard.__call__` (this is the `guard()` function). For more information on how to use metadata, see [How to use Metadata](/docs/hub/how_to_guides/metadata.md).
+
+```python
+guard = Guard.from_rail("my_railspec.rail")
+
+outcome = guard(
+    llm_api=openai.chat.completions.create,
+    model="gpt-3.5-turbo",
+    num_reasks=3,
+    metadata={
+        "filepaths": [
+            "./my_data/article1.txt",
+            "./my_data/article2.txt",
+        ]
+    }
+)
+```
+
+## Custom Validators
+
+If you need to perform a validation that is not currently supported by the hub, you can create your own custom validators.
+
+A custom validator can be as simple as a single function if you do not require addtional arguments:
+
+```py
+from typing import Dict
+from guardrails.validators import (
+    FailResult,
+    PassResult,
+    register_validator,
+    ValidationResult,
+)
+
+@register_validator(name="starts-with-a", data_type="string")
+def starts_with_a(value: str, metadata: Dict) -> ValidationResult:
+    if value.startswith("a"):
+        return PassResult()
+
+    return FailResult(
+        error_message=f"Value {value} does not start with a.",
+        fix_value="a" + value,
+    )
+```
+
+:::note
+A fix_value was supplied in the FailResult in the example above. This value represents a programmatic fix that can be applied to the output if `on_fail='fix'` is passed during validator initialization.
+:::
+
+If you need to perform more complex operations or require addtional arguments to perform the validation, then the validator can be specified as a class that inherits from our base Validator class:
+
+```py
+from typing import Callable, Dict, Optional
+from guardrails.validators import (
+    FailResult,
+    PassResult,
+    register_validator,
+    ValidationResult,
+    Validator,
+)
+
+@register_validator(name="starts-with", data_type="string")
+class StartsWith(Validator):
+    def __init__(self, prefix: str, on_fail: Optional[Callable] = None):
+        super().__init__(on_fail=on_fail, prefix=prefix)
+        self.prefix = prefix
+
+    def validate(self, value: str, metadata: Dict) -> ValidationResult:
+        if value.startswith(self.prefix):
+            return PassResult()
+
+        return FailResult(
+            error_message=f"Value {value} does not start with {self.prefix}.",
+            fix_value=self.prefix + value,  # To enable the "fix" option for on-fail
+        )
+```
+
+Custom validators must be defined before creating a `Guard` or `RAIL` spec in the code, 
+but otherwise can be used like built in validators. It can be used in a `RAIL` spec OR
+a `Pydantic` model like so:
+
```

Comment:
is there a code block missing here? i noticed this also is missing in the source?

Reply:
yup... time to go back through git history